components/raftstore-v2/src/operation/command/admin/mod.rs

```diff
@@ -118,7 +126,81 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
                 AdminCmdType::Split => Err(box_err!(
                     "Split is deprecated. Please use BatchSplit instead."
                 )),
-                AdminCmdType::BatchSplit => self.propose_split(ctx, req),
+                AdminCmdType::BatchSplit => {
+                    #[allow(clippy::question_mark)]
+                    if let Err(err) = validate_batch_split(req.get_admin_request(), self.region()) {
+                        Err(err)
+                    } else {
+                        // To reduce the impact of the expensive operation of `checkpoint` (it will
+                        // flush memtables of the rocksdb) in applying batch split, we split the
+                        // BatchSplit cmd into two phases:
+                        //
+                        // 1. Schedule flush memtable task so that the memtables of the rocksdb can
+                        // be flushed in advance in a way that will not block the normal raft
+                        // operations (`checkpoint` will still cause flush but it will be
+                        // significantly lightweight). At the same time, send flush memtable msgs to
+                        // the follower so that they can flush memtalbes in advance too.
+                        //
+                        // 2. When the task finishes, it will propose a batch split with
+                        // `SPLIT_SECOND_PHASE` flag.
+                        if !WriteBatchFlags::from_bits_truncate(req.get_header().get_flags())
+                            .contains(WriteBatchFlags::SPLIT_SECOND_PHASE)
+                        {
+                            if self.tablet_being_flushed() {
+                                return;
+                            }
+
+                            let region_id = self.region().get_id();
+                            self.set_tablet_being_flushed(true);
+                            info!(
+                                self.logger,
+                                "Schedule flush tablet";
+                            );
+                            if let Err(e) = ctx.schedulers.tablet_flush.schedule(
+                                crate::TabletFlushTask::TabletFlush {
+                                    region_id,
+                                    req: Some(req),
+                                    is_leader: true,
+                                    ch: Some(ch),
+                                },
+                            ) {
+                                error!(
+                                    self.logger,
+                                    "Fail to schedule flush task";
+                                    "err" => ?e,
+                                )
+                            }
+
+                            let peers = self.region().get_peers().to_vec();
+                            for p in peers {
```

Comment:
Does it needs to send it for all peer includes witness or tiflash?

Reply:
Good catch. I think filtering witness and non-voter is enough.